wit/guest.wit

```diff
@@ -1,10 +1,7 @@
-interface messaging-guest {
-    use messaging-types.{message, guest-configuration, error};
+interface guest {
+    use types.{message, guest-configuration, error};
 
-    /// Returns the list of channels (and extension metadata within guest-configuration) that 
-    /// this component should subscribe to and be handled by the subsequent handler within guest-configuration
-    configure: func() -> result<guest-configuration, error>;
-
-    /// Whenever this guest receives a message in one of the subscribed channels, the message is sent to this handler
+    /// Whenever this guest receives a message in one of the subscribed channels, the message is
+    /// sent to this handler
     handler: func(ms: list<message>) -> result<_, error>;
```

Comment:
Is wasi-messaging meant to be implemented by messaging systems with reliable (at-most- or at-least-once) delivery?  If `handler` returns an `error`, should the system assume that every `message` in the `list<message>` failed?  But what if some of the `message`s succeeded (finished their work and performed side effects)?  Are failed messages ever retried and, if so, does a component targeting `wasi:messaging` need to be idempotent and, if so, it probably needs some sort of message-id to do that?

Reply:
> Is wasi-messaging meant to be implemented by messaging systems with reliable (at-most- or at-least-once) delivery?

Yes.

> If `handler` returns an error, should the system assume that every message in the `list<message>` failed?

What do you mean by “failed“? Any error returned by the `handler` is a processing error and should be completely handled by the guest - like, if a handler got a message, that message has been received and it is up to the guest implementor to properly settle it with one of the following methods: https://github.com/WebAssembly/wasi-messaging/blob/c2a7421ad0aece25d11ab5056a8b98f18bedd64a/wit/consumer.wit#L25-L26

That said, I do like the idea of ID-ing messages - I just don‘t know if that is something widely supported at all, implementors might have to create their own implementations for managing IDs if not, which is something we should consider.

Reply:
I toyed with the idea of publishing messages and handling messages being a single `message` rather than a list of them. The thought was we could start with that and then add batching, which introduces the complexity mentioned by @lukewagner, as a later thing. Any thoughts?

Reply:
Ah, I see: so `complete-message`/`abandon-message` allow you to mark each individual message of a batch; makes sense.  (FWIW, if we `resource`-ify `message` (as discussed above), then we don't need to add any extra "message-id": the handle to the `message` identifies the message and so, e.g., `complete` and `abandon` could be methods on the `message`.)

Regarding batching: if we processed one `message` at a time, then I suppose the return value of the `handle` function could indicate whether the (single) `message` parameter was completed, abandoned, or there was a ephemeral error that wants a retry (which is I suppose is how the host would interpret a trap that kills the instance before it returns?).  That does seem like a simpler API, but maybe it incurs more per-message overhead, but maybe it's fine, at least as a first version?  Some data might help know: does anyone know, for existing message-passing systems, how long it takes to process the average message?  (If it's longer than, say, 1 microsecond, then batching may not be a significant win given the speed of host-to-component calls).

Reply:
FWIW we have several people using a custom messaging interface we built for wasmCloud (until this one is ready) in production. It handles and publishes single messages: https://github.com/wasmCloud/messaging/blob/main/wit/messaging.wit. I think it might be worth having batching be a separate interface we define further down the road